File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js


function solution(maps) {
    const ylen = maps.length
    const xlen = maps[0].length
    const goalx = xlen-1
    const goaly = ylen -1
    const dx = [0,0,1,-1]
    const dy = [-1,1,0,0]
    const queue = []
    queue.push([0,0,1])
    while(queue.length) {
        const [cury, curx, move] = queue.shift()
        if(curx === goalx && cury === goaly) {
            return move
        }
        for(let i = 0; i<4; i++) {
            const nx = curx+dx[i]
            const ny = cury+dy[i]
            if(nx >= 0 && ny >= 0 && nx < xlen && ny < ylen && maps[ny][nx] === 1){
                queue.push([ny,nx, move+1])
                maps[ny][nx] = 0
            }
        }
    }
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

dx = [1, -1, 0, 0]
dy = [0, 0, -1, 1]

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    visited = [[0]*m for _ in range(n)]
    
    def bfs(x, y):
        queue = deque()
        queue.append((x, y))
        visited[x][y] = 1
        
        while queue:
            x, y = queue.popleft()
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if 0 <= nx < n and 0 <= ny < m:
                    if visited[nx][ny] == 0 and maps[nx][ny] == 1:
                        visited[nx][ny] = visited[x][y] + 1
                        queue.append((nx, ny))
    bfs(0, 0)
    if visited[n-1][m-1] == 0:
        answer = -1
    else:
        answer = visited[n-1][m-1]
    return answer